Add "find references" capability to Ivy integrated Language Service

atscott opened the pull request:
Commit 1:
In order to map the pipe's `transform` method in the type check block
directly back to the pipe name in the template source, we need to
include the `BindingPipe`'s `nameSpan` with the `ts.methodAccess` for
the pipe's transform method.

Note that this is specifically relevant to the Language Service's "find
references" feature. As an example, with something like `-2.5 | number:'1.0-0'`,,
when calling "find references" on the 'number' pipe we want the text
span of the reference to just be `number` rather than the entire binding
pipe's source `-2.5 | number:'1.0-0'`.

Commit 2:
This commit adds "find references" functionality to the Ivy integrated
language service. The basic approach is as follows:

1. Generate shims for all files to ensure we find references in shims
throughout the entire program
2. Determine if the position for the reference request is within a
template.
  * Yes, it is in a template: Find which node in the template AST the
  position refers to. Then find the position in the shim file for that
  template node. Pass the shim file and position in the shim file along
  to step 3.
  * No, the request for references was made outside a template: Forward
  the file and position to step 3.
3. (`getReferencesAtTypescriptPosition`): Call the native TypeScript LS
`getReferencesAtPosition`. For each reference that is in a shim file, map those
back to a template location, otherwise return it as-is.

Close https://github.com/angular/vscode-ng-language-service/issues/29

atscott committed:
feat(language-service): Add "find references" capability to Ivy integrated LS

This commit adds "find references" functionality to the Ivy integrated
language service. The basic approach is as follows:

1. Generate shims for all files to ensure we find references in shims
throughout the entire program
2. Determine if the position for the reference request is within a
template.
  * Yes, it is in a template: Find which node in the template AST the
  position refers to. Then find the position in the shim file for that
  template node. Pass the shim file and position in the shim file along
  to step 3.
  * No, the request for references was made outside a template: Forward
  the file and position to step 3.
3. (`getReferencesAtTypescriptPosition`): Call the native TypeScript LS
`getReferencesAtPosition`. For each reference that is in a shim file, map those
back to a template location, otherwise return it as-is.

atscott committed:
refactor(compiler-cli): Include pipe `nameSpan` in TCB

In order to map the pipe's `transform` method in the type check block
directly back to the pipe name in the template source, we need to
include the `BindingPipe`'s `nameSpan` with the `ts.methodAccess` for
the pipe's transform method.

Note that this is specifically relevant to the Language Service's "find
references" feature. As an example, with something like `-2.5 | number:'1.0-0'`,,
when calling "find references" on the 'number' pipe we want the text
span of the reference to just be `number` rather than the entire binding
pipe's source `-2.5 | number:'1.0-0'`.

ayazhafiz commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,151 @@
+
+    const entries: ts.ReferenceEntry[] = [];
+    for (const ref of refs) {
+      if (ref.fileName.endsWith('ngtypecheck.ts')) {
```
maybe we should expose a `isShimFilePath` on the type check strategy? Can be a todo.

kyliau replied:
We do have such a method, but you'll have to convert the filename to a source file first:
https://github.com/angular/angular/blob/23c36a24ede7483a032f19a186ec29f7301768eb/packages/language-service/ivy/adapters.ts#L33-L35

atscott replied:
@kyliau - Alex, pointed me to the `isShim` helper as well, but it doesn't work because we store the data on a specific instance of the `SourceFile` object. We can't do `isShim(program.getSourceFile(fileName))` because that returns a different object, without the shim marker.

One potential solution would be to expose an `isShim` method on the `TemplateTypeChecker` that makes use of the `getFileAndShimRecordsForPath` method and returns `true` if it's able to find the file in its records.

// cc @alxhub

alxhub replied:
`isShim` won't work because the Language Service generates shims in a completely different way than the shim generator that `isShim` is designed for.

ayazhafiz commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,151 @@
+          entries.push(entry);
+        }
+      } else {
+        // If the request for a reference is from a template, we want to include regular TS file
+        // results because the TS LS will not find those from the template string/external template.
+        // Otherwise, we skip them because the TS LS *will* include those in the results.
+        if (requestMadeFromTemplate) {
+          entries.push(ref);
+        }
```
I think the editor should be doing this, not us. In my mind TS references are still in the scope of an Angular app, and relying on TS to give them back as supplemental assumes a TS LS is enabled simultaneously as ours, which is reasonable but not something we should rely on IMO. What do you think?

ayazhafiz replied:
How about we de-duplicate on the language server (extension) side? This way we are more "near the editor", which ideally should be responsible for this, the language service itself does not have to be concerned with this. And we can upstream a bug report to vscode folks as well to see what they think. Curious if @kyliau @alxhub have thoughts too.

ayazhafiz replied:
Update: I looked into this more because I remembered I had some vague recollection of how this worked, kind of like some itch bugging me. Eventually I found https://github.com/microsoft/vscode/issues/39890 which I saw long ago, which suggests that vscode should deduplicate "identical" results. I then looked at VSCode's reference provider code, [found here](https://github.com/microsoft/vscode/blob/ea4d99921cc790d49194e897021faee02a1847f7/src/vs/editor/contrib/gotoSymbol/referencesModel.ts#L144-L174), and it looks like it considers [equal references those in the same file and the same span](https://github.com/microsoft/vscode/blob/ea4d99921cc790d49194e897021faee02a1847f7/src/vs/editor/contrib/gotoSymbol/referencesModel.ts#L283-L285). So it seems that if VSCode is not deduplicating results this should be a bug upstream, as it seems it is their intention to do so.

ayazhafiz commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,151 @@
+    templateTypeChecker: TemplateTypeChecker): ts.ReferenceEntry|null {
+  const mapping = templateTypeChecker.getTemplateMappingAtShimLocation({
+    shimPath: absoluteFrom(shimReferenceEntry.fileName),
+    positionInShimFile: shimReferenceEntry.textSpan.start
```
nit: add a comma at the end of this line

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+  });
+
+  it('getting references from app component', () => {
+    const fileWithCursor = getBasicAppFileContents({
+      classContents: 'myP¦rop: string',
+      componentMeta: {templateUrl: 'app.html'},
+    });
```
This is a little difficult for me to parse when reading tests, could we type `getBasicAppFileContents` as `getBasicAppFileContents(clazz: string)` and pass the entire string of what the class looks like? I realize it's kind of a pain to rewrite everything you already wrote, so I'm not strongly for this, but just a thought.

atscott replied:
Yea, I wasn't totally happy with how the tests were reading. Working on it now - I just pushed a change to `getBasicAppFileContents`, renaming it to `createModuleWithDeclarations`

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+    initMockFileSystem('Native');
+  });
+
+  it('getting references from app component', () => {
```
nit: `getting references from app component -> gets component member references from TS file`

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+    expect(result[0].textSpan).toEqual('myProp');
+  });
+
+  it('getting references from app component', () => {
```
nit: `gets component member references from template`

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+        assertTextSpans(refs, ['dirRef']);
+      });
+
+      it('should work for prop reads of directive references', () => {
```
what about safe prop reads? (kind of esoteric but I don't think it hurts to add)

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+        }
+      `;
+
+      function generateAppFile(template: string) {
+        return `
+        import {Component, NgModule} from '@angular/core';
+        import {Dir} from './dir';
+
+        @Component({template: '${template}'})
+        export class AppCmp {}
+
+        @NgModule({declarations: [AppCmp, Dir]})
+        export class AppModule {}`;
+      }
```
This seems very similar to `getBasicAppFileContents`; could we scope and name the two functions with respect to the kinds of specs they are used for, or use one or the other?

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+      assertTextSpans(refs, ['iRef']);
+    });
+
+    it('should work for original name of assigned variable', () => {
```
I think a better name for this spec is `should work for initializer of variable`; what do you think?

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+  });
+
+  describe('pipes', () => {
+    it('should work for pipes', () => {
```
nit: `should work for pipe names`

Can you add specs for pipe arguments?

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+      assertTextSpans(refs, ['model']);
+    });
+
+    it('should work from the input', () => {
```
nit: `should work from the TS input declaration`

Can you also add a spec of usage of the input in a TS script, aside from its declaration? (If that works)

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+
+  function getReferencesAtPosition(fileName: string, position: number) {
+    const result = env.ngLS.getReferencesAtPosition(fileName, position);
+    return result === undefined ? undefined : result.map(humanizeReferenceEntry);
```
`return result?.map(humanizeReferenceEntry); `

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+    return result === undefined ? undefined : result.map(humanizeReferenceEntry);
+  }
+
+  function humanizeReferenceEntry(entry: ts.ReferenceEntry) {
```
Can we type the return ty of this function?

ayazhafiz replied:
Oh man, I assumed the type was something standard in the ts compiler API.... now I'm sorry for making you do that. anyway, it is my favorite return type in this pr tbh ;)

ayazhafiz reviewed (commented):
This is very nice!

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+      const fileWithCursor = getBasicAppFileContents({
+        classContents: `heroes: string[] = []`,
+        componentMeta:
+            {template: `<div *ngFor="let hero of heroes; let iRef = index">{{iR¦ef}}</div>`},
+      });
+      const {text, cursor} = extractCursorInfo(fileWithCursor);
+      env = LanguageServiceTestEnvironment.setup([
+        {name: _('/app.ts'), contents: text, isRoot: true},
+      ]);
+      const refs = getReferencesAtPosition(_('/app.ts'), cursor)!;
+      expect(refs.length).toBe(2);
+      assertFileNames(refs, ['app.ts']);
+      assertTextSpans(refs, ['iRef']);
```
Can you validate the first `iRef` is the definition reference?

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+    it('should work for implicit variables', () => {
+      const fileWithCursor = getBasicAppFileContents({
+        classContents: `heroes: string[] = []`,
+        componentMeta: {template: `<div *ngFor="let hero of heroes">{{her¦o}}</div>`},
+      });
+      const {text, cursor} = extractCursorInfo(fileWithCursor);
+      env = LanguageServiceTestEnvironment.setup([
+        {name: _('/app.ts'), contents: text, isRoot: true},
+      ]);
+      const refs = getReferencesAtPosition(_('/app.ts'), cursor)!;
+      expect(refs.length).toBe(2);
+      assertFileNames(refs, ['app.ts']);
+      assertTextSpans(refs, ['hero']);
```
Can you validate the first `hero` is the definition reference?

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+        export class AppModule {}`;
+      }
+      it('should work', () => {
+        const fileWithCursor = generateAppFile('<div [dir] #dirRef="myDir"></div> {{ dirR¦ef }}');
+        const {text, cursor} = extractCursorInfo(fileWithCursor);
+        env = LanguageServiceTestEnvironment.setup([
+          {name: _('/app.ts'), contents: text, isRoot: true},
+          {name: _('/dir.ts'), contents: dirFile},
+        ]);
+        const refs = getReferencesAtPosition(_('/app.ts'), cursor)!;
+        expect(refs.length).toBe(2);
+        assertFileNames(refs, ['app.ts']);
+        assertTextSpans(refs, ['dirRef']);
```
Can you validate the first `dirRef` is the definition reference?

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,514 @@
+  });
+
+  describe('references', () => {
+    it('should work', () => {
```
```suggestion
    it('should work for element references', () => {
```
Could we add a test for template references?

atscott committed:
refactor(compiler-cli): Add nameSpan to SafePropertyRead TCB

In order to map the a safe property read's method access in the type check block
directly back to the property in the template source, we need to
include the `SafePropertyRead`'s `nameSpan` with the `ts.propertyAccess` for
the pipe's transform method.

Note that this is specifically relevant to the Language Service's "find
references" feature. As an example, with something like `{{a?.value}}`,
when calling "find references" on the 'value' we want the text
span of the reference to just be `value` rather than the entire source
`a?.value`.

atscott reviewed (commented):
> Can you also add a spec of usage of the input in a TS script, aside from its declaration? (If that works)

I think I understood what you meant and added a test, but I'm not entirely sure. See `should work for inputs referenced from some other place`

ayazhafiz commented on packages/compiler-cli/src/ngtsc/testing/fake_common/index.ts:
```diff
@@ -11,6 +11,12 @@ import {NgIterable, TemplateRef, ɵɵDirectiveDefWithMeta, ɵɵNgModuleDefWithMe
 export interface NgForOfContext<T, U extends NgIterable<T>> {
   $implicit: T;
   ngForOf: U;
+  odd: boolean;
```
Are the language service tests plugging into the compiler tests?

ayazhafiz commented on packages/language-service/ivy/language_service.ts:
```diff
@@ -97,6 +97,14 @@ export class LanguageService {
   }
 
+  getReferencesAtPosition(fileName: string, position: number): ts.ReferenceEntry[]|undefined {
+    const compiler = this.compilerFactory.getOrCreateWithChangedFile(fileName);
+    const results = new ReferenceBuilder(this.strategy, this.tsLS, compiler)
+                        .get(absoluteFrom(fileName), position);
+    this.compilerFactory.registerLastKnownProgram();
+    return results;
```
aside, no action required: we should create a `withCompiler` or similar wrapper in this class, having to make sure we register the last known program is making me shiver a bit

ayazhafiz commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,146 @@
+    shimReferenceEntry: ts.ReferenceEntry,
+    templateTypeChecker: TemplateTypeChecker): ts.ReferenceEntry|null {
+  const mapping = templateTypeChecker.getTemplateMappingAtShimLocation({
+    shimPath: absoluteFrom(shimReferenceEntry.fileName),
```
Here and on line 130 and 132: consider using the `resolve` method on the `project` `serverHost` rather than the compiler filesystem. `LSParseConfigHost` in the adapters file provides this functionality as well. As discussed before this generally should not be a problem, but I think it would be good to be consistent in what abstraction we use to  resolve paths.

atscott replied:
@alxhub - Can you provide your input here. I only added the `absoluteFrom` after a nudge from you to do it. Otherwise, these would just pass around the string `fileName`s.

atscott replied:
Added TODO after discussion with Alex. I think `absoluteFrom` is fine for now since we haven't really considered or tested remote files at all. We should be resolving to some sort of branded path before passing a path along to the compiler and I'm not really sure what that would look like at the moment.

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+          @Component({templateUrl: './app.html' })
+          export class AppCmp {
+            title = '';
+          }`
```
very nit, here and other places: 
```suggestion
          }`,
```

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+      createModuleWithDeclarations([appFile]);
+      const refs = getReferencesAtPosition(_('/app.ts'), cursor)!;
+      expect(refs.length).toBe(2);
+      assertTextSpans(refs, ['myInput']);
```
In all the places where `assertTextSpans` is called, it appears the array is a singleton. I think this check could be more explicit by requiring `assertTextSpans` to only accept one string that must match for all the references, or have the array consist of the same # of items as the # of references, each corresponding to that reference's span. What do you think?

atscott replied:
There's only one string _most_ of the time, but there are a few cases where the text spans have different texts: pipes, alias inputs/outputs, and directives. There's also no limit to the number of results returned, nor the order. I didn't find a ton of value in a more restrictive assertion like `assertTextSpans(['Dir', 'Dir', 'Dir', '<div dir>'])` or `assertTextSpans(['myProp', 'myProp'])` given the assertion on the number of results. As long as the number of results are correct and the spans are in the set of expected results, I can't think of a scenario where these two assertions combined would fail.

ayazhafiz replied:
sgtm, I don't think it's a big deal either way

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+  });
+
+  describe('variables', () => {
+    it('should work for implicit variables', () => {
```
```suggestion
    it('should work for variable initialized implicitly', () => {
```

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+      });
+
+      it('should work for safe method calls', () => {
+        const fileWithCursor = '<div [dir] #dirRef="myDir"></div> {{ dirRef?.doSometh¦ing }}';
```
```suggestion
        const fileWithCursor = '<div [dir] #dirRef="myDir"></div> {{ dirRef?.doSometh¦ing() }}';
```

ayazhafiz commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+    assertTextSpans(refs, ['title']);
+  });
+
+  it('should work for property writes', () => {
```
Some more test cases we may want to cover:

- references for the rvalue of a property write
- references for constant keyed reads (for example, `{{a["b"]}}` where `a = {b: string}` should work as it works in TS)

atscott replied:
Done. These two tests actually brought up two issues:
1. rvalue of property write didn't work because the RHS wasn't wrapped in parens again if it was already a parenthesized expression. This was wrong because the existing parenthesized expression had a trailing comment, which then matched to the whole assignment node
2. We currently return the whole string literal span for keyed reads. I added a comment to the test to investigate updating this in the future in a way similar to how the TSLS does it.

ayazhafiz replied:
good that you added them then :)

kyliau commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,146 @@
+      }
+      case SymbolKind.Input:
+      case SymbolKind.Output: {
+        // TODO(atscott): Determine how to handle when the binding maps to several inputs/outputs
```
We need to take into account two way bindings too, which I think at this point have been converted to separate inputs and outputs. Could you please confirm, and if so, add a TODO?

atscott replied:
The need to account for the two-way binding is part of the template target: https://github.com/angular/angular/blob/5f06438f7a15676dba7013f58daac10b77649a51/packages/language-service/ivy/template_target.ts#L169. Since the template target only returns one result, there's not exactly any special accounting that needs to be done downstream here. This algorithm assumes that the target is just the `BoundAttribute`. 

If you'd still like me to add a TODO here as well, I can do that.

kyliau replied:
SGTM, I've created a new card to keep track of this issue: https://github.com/orgs/angular/projects/1#card-50332681

kyliau commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,146 @@
+  private getReferencesAtTypescriptPosition(fileName: string, position: number):
+      ts.ReferenceEntry[]|undefined {
+    const refs = this.tsLS.getReferencesAtPosition(fileName, position);
+    if (!refs) return;
```
nit: prefer explicit check

kyliau commented on packages/language-service/ivy/language_service.ts:
```diff
@@ -97,6 +97,14 @@ export class LanguageService {
     return results;
   }
 
+  getReferencesAtPosition(fileName: string, position: number): ts.ReferenceEntry[]|undefined {
```
Side question: Do you know what is the difference between `tsLS.getReferencesAtPosition()` and `tsLS.findReferences()`?

kyliau commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+    const {text, cursor} = extractCursorInfo(`
+          import {Component} from '@angular/core';
+
+          @Component({templateUrl: './app.html'})
```
Could you please add a similar test, but using inline template here?
We should make sure we can get references from TS to inline template.

atscott replied:
There are several tests that are testing inline templates. Is that enough or do you want to test every case with both inline and external templates?

kyliau replied:
I meant this test specifically -> initiate a `getReferences()` operation from TS code (not inline template), and make sure the result list includes symbol in inline template. I don't see such a test here, but if I missed it, please resolve this conversation.

atscott replied:
Ah, I understand now. Done.

kyliau commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,146 @@
+          entries.push(entry);
+        }
+      } else {
+        entries.push(ref);
```
Should we handle the case where get references is initiated from TS code (not Angular template), so TS references are already handled by the native TypeScript extension?
I saw we already had a discussion on this, but if we could reduce the result set sent from the server without too much complexity, then we should do that.

atscott replied:
Yea...so we should just decide what we want to do here. I had the logic to exclude the result but removed it after this discussion: https://github.com/angular/angular/pull/39768#discussion_r527289000. It's not difficult to manage either way, but we do need to make a call.

kyliau commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+
+      const originalRefs = env.ngLS.getReferencesAtPosition(_('/app.ts'), cursor)!;
+      // Get the declaration by finding the reference that appears first in the template
+      originalRefs.sort((a, b) => a.textSpan.start - b.textSpan.start);
```
Should we sort the result in the implementation?

ayazhafiz replied:
I don't think we should, consumers should not expect the references to be sorted anyway.

kyliau commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,674 @@
+          @Output('alias') aliasedModelChange = new EventEmitter<string>();
+        }`;
+
+    function generateAppFile(template: string) {
```
super-nit: consider moving this to top-level.

atscott replied:
This one is specific to the outputs tests which use the `StringModel` directive defined here.

atscott commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,151 @@
+      case SymbolKind.Element:
+      case SymbolKind.Template:
+      case SymbolKind.DomBinding:
+        // References to elements, templates, and directives will be through template references
+        // (#ref). They shouldn't be used directly for a Language Service reference request.
```
I think the only one that makes sense in the list here is potentially the `Element`, and only if it's a component. Potentially if the position is on an attribute which matches a directive selector too. I'll add a TODO here for this, but I don't think it should be a blocker for the initial implementation.

A `Directive` symbol is only included here for completeness of the switch case - it's never returned as a top-level symbol, but only as part of the directives list in the `Element` symbol.  `Template` symbols don't really have references unless there is a proper local ref (`#ref`) for them.

ayazhafiz replied:
Agreed that we don't need to do this right now. I think that maybe there are some design questions we can have here, but we can do that some other time

atscott committed:
refactor(compiler-cli): Always wrap RHS of TCB writes in parens

There were two issues with the current TCB:

1. The logic for only wrapping the right hand side of the property write
if it was not already a parenthesized expression was incorrect. A
parenthesized expression could still have a trailing comment, and if
that were the case, that span comment would still be ambiguous, as explained
by the comment in the code before `wrapForTypeChecker`.
2. The right hand side of keyed writes was not wrapped in parens at all

alxhub commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,151 @@
+      }
+      case SymbolKind.Expression: {
+        const {shimPath, positionInShimFile} = symbol.shimLocation;
+        return this.getReferencesAtTypescriptPosition(shimPath, positionInShimFile);
```
It might be valuable to validate that the expression here is actually of a kind that makes sense to search for references on - a `PropertyRead` or `MethodCall`?

atscott replied:
Hmm, I think it might be better to just make the request and allow the TSLS to just return no results. In addition to `PropertyRead` and `MethodCall`, there is also `BindingPipe` (though this would change to its own symbol type after #39555) and `LiteralPrimitive` for keyed reads (`something['a']`). I could go both ways, but I'm leaning towards keeping this as-is to avoid the extra logic where it will already return no results (as far as I know).

alxhub commented on packages/language-service/ivy/references.ts:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  return {
+    isWriteAccess: true,
```
Copy this from the original entry - it refers to whether the reference in question is directly assigned to.

alxhub commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,760 @@
+    const templateFile = {name: _('/app.html'), contents: text};
+    createModuleWithDeclarations([appFile], [templateFile]);
+    const refs = getReferencesAtPosition(_('/app.html'), cursor)!;
+    // Angular LS returns the template ref _and_ component member because TSLS doesn't know how to
```
Delete comment

alxhub commented on packages/language-service/ivy/test/references_spec.ts:
```diff
@@ -0,0 +1,760 @@
+        dirFile = {name: _('/dir.ts'), contents: dirFileContents};
+      });
+
+      it('should work', () => {
```
Better test name?

atscott committed:
refactor(compiler-cli): expose TTC method to determine if file is tracked shim

The Language Service "find references" currently uses the
`ngtypecheck.ts` suffix to determine if a file is a shim file. Instead,
a better API would be to expose a method in the template type checker
that does this verification so that the LS does not have to "know" about
the typecheck suffix. This also fixes an issue (albeit unlikely) whereby a file
in the user's program that _actually_ is named with the `ngtypecheck.ts`
suffix would have been interpreted as a shim file.

jessicajaniuk commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:345107485:BASE:345107506:1606863697891:1481a0b9)
